Index: sys/opencrypto/cryptosoft.c
===================================================================
--- sys/opencrypto/cryptosoft.c	(revision 340349)
+++ sys/opencrypto/cryptosoft.c	(working copy)
@@ -812,6 +812,12 @@
 		case CRYPTO_CHACHA20:
 			txf = &enc_xform_chacha20;
 			goto enccommon;
+        case CRYPTO_THREEFISH_CBC:
+            txf = &enc_xform_threefish;
+            goto enccommon;
+        case CRYPTO_THREEFISH_XTS:
+            txf = &enc_xform_threefish_xts;
+            goto enccommon;
 		enccommon:
 			if (cri->cri_key != NULL) {
 				error = txf->setkey(&swd->sw_kschedule,
@@ -1042,6 +1048,8 @@
 		case CRYPTO_CAMELLIA_CBC:
 		case CRYPTO_NULL_CBC:
 		case CRYPTO_CHACHA20:
+        case CRYPTO_THREEFISH_CBC:
+        case CRYPTO_THREEFISH_XTS:
 			txf = swd->sw_exf;
 
 			if (swd->sw_kschedule)
@@ -1162,6 +1170,8 @@
 		case CRYPTO_AES_ICM:
 		case CRYPTO_CAMELLIA_CBC:
 		case CRYPTO_CHACHA20:
+        case CRYPTO_THREEFISH_CBC:
+        case CRYPTO_THREEFISH_XTS:
 			if ((crp->crp_etype = swcr_encdec(crd, sw,
 			    crp->crp_buf, crp->crp_flags)) != 0)
 				goto done;
@@ -1289,6 +1299,8 @@
 	REGISTER(CRYPTO_BLAKE2S);
 	REGISTER(CRYPTO_CHACHA20);
 	REGISTER(CRYPTO_POLY1305);
+    REGISTER(CRYPTO_THREEFISH_CBC);
+    REGISTER(CRYPTO_THREEFISH_XTS);
 #undef REGISTER
 
 	return 0;
