Index: sys/opencrypto/cryptodev.h
===================================================================
--- sys/opencrypto/cryptodev.h	(revision 345236)
+++ sys/opencrypto/cryptodev.h	(working copy)
@@ -127,7 +127,8 @@
 #define	ARC4_BLOCK_LEN		1
 #define	CAMELLIA_BLOCK_LEN	16
 #define	CHACHA20_NATIVE_BLOCK_LEN	64
-#define	EALG_MAX_BLOCK_LEN	CHACHA20_NATIVE_BLOCK_LEN /* Keep this updated */
+#define THREEFISH_BLOCK_LEN 64
+#define	EALG_MAX_BLOCK_LEN	THREEFISH_BLOCK_LEN /* Keep this updated */
 
 /* IV Lengths */
 
@@ -160,6 +161,8 @@
 #define	ARC4_MAX_KEY		32
 #define	CAMELLIA_MIN_KEY	8
 #define	CAMELLIA_MAX_KEY	32
+#define THREEFISH_MIN_KEY   64
+#define THREEFISH_MAX_KEY   THREEFISH_MIN_KEY
 
 /* Maximum hash algorithm result length */
 #define	AALG_MAX_RESULT_LEN	64 /* Keep this updated */
@@ -206,7 +209,9 @@
 #define	CRYPTO_POLY1305		38
 #define	CRYPTO_AES_CCM_CBC_MAC	39	/* auth side */
 #define	CRYPTO_AES_CCM_16	40	/* cipher side */
-#define	CRYPTO_ALGORITHM_MAX	40	/* Keep updated - see below */
+#define CRYPTO_THREEFISH_CBC    41
+#define CRYPTO_THREEFISH_XTS    42
+#define	CRYPTO_ALGORITHM_MAX	42	/* Keep updated - see below */
 
 #define	CRYPTO_ALGO_VALID(x)	((x) >= CRYPTO_ALGORITHM_MIN && \
 				 (x) <= CRYPTO_ALGORITHM_MAX)
