Index: sys/opencrypto/cryptosoft.c
===================================================================
--- sys/opencrypto/cryptosoft.c	(revision 345236)
+++ sys/opencrypto/cryptosoft.c	(working copy)
@@ -870,6 +870,12 @@
 		case CRYPTO_CHACHA20:
 			txf = &enc_xform_chacha20;
 			goto enccommon;
+        case CRYPTO_THREEFISH_CBC:
+            txf = &enc_xform_threefish;
+            goto enccommon;
+        case CRYPTO_THREEFISH_XTS:
+            txf = &enc_xform_threefish_xts;
+            goto enccommon;
 		enccommon:
 			if (cri->cri_key != NULL) {
 				error = txf->setkey(&swd->sw_kschedule,
@@ -1117,6 +1123,8 @@
 		case CRYPTO_NULL_CBC:
 		case CRYPTO_CHACHA20:
 		case CRYPTO_AES_CCM_16:
+        case CRYPTO_THREEFISH_CBC:
+        case CRYPTO_THREEFISH_XTS:
 			txf = swd->sw_exf;
 
 			if (swd->sw_kschedule)
@@ -1241,6 +1249,8 @@
 		case CRYPTO_AES_ICM:
 		case CRYPTO_CAMELLIA_CBC:
 		case CRYPTO_CHACHA20:
+        case CRYPTO_THREEFISH_CBC:
+        case CRYPTO_THREEFISH_XTS:
 			if ((crp->crp_etype = swcr_encdec(crd, sw,
 			    crp->crp_buf, crp->crp_flags)) != 0)
 				goto done;
@@ -1372,6 +1382,8 @@
 	REGISTER(CRYPTO_AES_CCM_16);
 	REGISTER(CRYPTO_AES_CCM_CBC_MAC);
 	REGISTER(CRYPTO_POLY1305);
+    REGISTER(CRYPTO_THREEFISH_CBC);
+    REGISTER(CRYPTO_THREEFISH_XTS);
 #undef REGISTER
 
 	return 0;
