Index: sys/opencrypto/cryptodev.c
===================================================================
--- sys/opencrypto/cryptodev.c	(revision 345236)
+++ sys/opencrypto/cryptodev.c	(working copy)
@@ -447,6 +447,12 @@
 		case CRYPTO_AES_CCM_16:
 			txform = &enc_xform_ccm;
 			break;
+        case CRYPTO_THREEFISH_CBC:
+            txform = &enc_xform_threefish;
+            break;
+        case CRYPTO_THREEFISH_XTS:
+            txform = &enc_xform_threefish_xts;
+            break;
 
 		default:
 			CRYPTDEB("invalid cipher");
