Index: sys/opencrypto/cryptosoft.c
===================================================================
--- sys/opencrypto/cryptosoft.c	(revision 320320)
+++ sys/opencrypto/cryptosoft.c	(working copy)
@@ -802,6 +802,12 @@
 		case CRYPTO_CAMELLIA_CBC:
 			txf = &enc_xform_camellia;
 			goto enccommon;
+		case CRYPTO_THREEFISH_CBC:
+			txf = &enc_xform_threefish;
+			goto enccommon;
+		case CRYPTO_THREEFISH_XTS:
+			txf = &enc_xform_threefish_xts;
+			goto enccommon;
 		case CRYPTO_NULL_CBC:
 			txf = &enc_xform_null;
 			goto enccommon;
@@ -1012,6 +1018,8 @@
 		case CRYPTO_AES_NIST_GCM_16:
 		case CRYPTO_AES_NIST_GMAC:
 		case CRYPTO_CAMELLIA_CBC:
+		case CRYPTO_THREEFISH_CBC:
+		case CRYPTO_THREEFISH_XTS:
 		case CRYPTO_NULL_CBC:
 			txf = swd->sw_exf;
 
@@ -1138,6 +1146,8 @@
 		case CRYPTO_AES_XTS:
 		case CRYPTO_AES_ICM:
 		case CRYPTO_CAMELLIA_CBC:
+		case CRYPTO_THREEFISH_CBC:
+		case CRYPTO_THREEFISH_XTS:
 			if ((crp->crp_etype = swcr_encdec(crd, sw,
 			    crp->crp_buf, crp->crp_flags)) != 0)
 				goto done;
@@ -1246,6 +1256,8 @@
 	REGISTER(CRYPTO_AES_192_NIST_GMAC);
 	REGISTER(CRYPTO_AES_256_NIST_GMAC);
  	REGISTER(CRYPTO_CAMELLIA_CBC);
+ 	REGISTER(CRYPTO_THREEFISH_CBC);
+ 	REGISTER(CRYPTO_THREEFISH_XTS);
 	REGISTER(CRYPTO_DEFLATE_COMP);
 #undef REGISTER
 
