Index: sys/opencrypto/xform_enc.h
===================================================================
--- sys/opencrypto/xform_enc.h	(revision 345236)
+++ sys/opencrypto/xform_enc.h	(working copy)
@@ -37,6 +37,7 @@
 #include <crypto/des/des.h>
 #include <crypto/rijndael/rijndael.h>
 #include <crypto/camellia/camellia.h>
+#include <crypto/threefish/threefish.h>
 #include <opencrypto/cast.h>
 #include <opencrypto/skipjack.h>
 #include <opencrypto/cryptodev.h>
@@ -85,6 +86,8 @@
 extern struct enc_xform enc_xform_camellia;
 extern struct enc_xform enc_xform_chacha20;
 extern struct enc_xform enc_xform_ccm;
+extern struct enc_xform enc_xform_threefish;
+extern struct enc_xform enc_xform_threefish_xts;
 
 struct aes_icm_ctx {
 	u_int32_t	ac_ek[4*(RIJNDAEL_MAXNR + 1)];
@@ -99,4 +102,9 @@
 	u_int8_t tweak[AES_XTS_BLOCKSIZE];
 };
 
+struct threefish_xts_ctx {
+    threefish_ctx st;
+    u_int64_t tweak[2];
+}
+
 #endif /* _CRYPTO_XFORM_ENC_H_ */
