Index: share/mk/bsd.libnames.mk
===================================================================
--- share/mk/bsd.libnames.mk	(revision 313496)
+++ share/mk/bsd.libnames.mk	(working copy)
@@ -146,6 +146,7 @@
 LIBTERMCAPW?=	${DESTDIR}${LIBDIR}/libtermcapw.a
 LIBTERMLIB?=	"don't use LIBTERMLIB, use LIBTERMCAP"
 LIBTINFO?=	"don't use LIBTINFO, use LIBNCURSES"
+LIBTHREEFISH?=	${DESTDIR}${LIBDIR}/libthreefish.a
 LIBUFS?=	${DESTDIR}${LIBDIR}/libufs.a
 LIBUGIDFW?=	${DESTDIR}${LIBDIR}/libugidfw.a
 LIBULOG?=	${DESTDIR}${LIBDIR}/libulog.a
