Index: share/mk/src.libnames.mk
===================================================================
--- share/mk/src.libnames.mk	(revision 340349)
+++ share/mk/src.libnames.mk	(working copy)
@@ -168,6 +168,7 @@
 		tacplus \
 		termcap \
 		termcapw \
+        threefish \
 		ufs \
 		ugidfw \
 		ulog \
@@ -560,6 +561,7 @@
 LIBPANELDIR=	${OBJTOP}/lib/ncurses/panel
 LIBPANELWDIR=	${OBJTOP}/lib/ncurses/panelw
 LIBCRYPTODIR=	${OBJTOP}/secure/lib/libcrypto
+LIBTHREEFISHDIR=${OBJTOP}/secure/lib/libthreefish
 LIBSSHDIR=	${OBJTOP}/secure/lib/libssh
 LIBSSLDIR=	${OBJTOP}/secure/lib/libssl
 LIBTEKENDIR=	${OBJTOP}/sys/teken/libteken
