Index: share/mk/src.libnames.mk
===================================================================
--- share/mk/src.libnames.mk	(revision 332871)
+++ share/mk/src.libnames.mk	(working copy)
@@ -161,6 +161,7 @@
 		tacplus \
 		termcap \
 		termcapw \
+		threefish \
 		ufs \
 		ugidfw \
 		ulog \
@@ -539,6 +540,7 @@
 LIBPANELDIR=	${OBJTOP}/lib/ncurses/panel
 LIBPANELWDIR=	${OBJTOP}/lib/ncurses/panelw
 LIBCRYPTODIR=	${OBJTOP}/secure/lib/libcrypto
+LIBTHREEFISHDIR=${OBJTOP}/secure/lib/libthreefish
 LIBSSHDIR=	${OBJTOP}/secure/lib/libssh
 LIBSSLDIR=	${OBJTOP}/secure/lib/libssl
 LIBTEKENDIR=	${OBJTOP}/sys/teken/libteken
