Index: sys/modules/crypto/Makefile
===================================================================
--- sys/modules/crypto/Makefile	(revision 320320)
+++ sys/modules/crypto/Makefile	(working copy)
@@ -19,6 +19,12 @@
 SRCS	+= des_ecb.c des_enc.c des_setkey.c
 SRCS	+= sha1.c sha256c.c sha512c.c
 SRCS	+= skein.c skein_block.c
+.if exists(${MACHINE_ARCH}/threefish_512.S)
+.PATH:	${.CURDIR}/../../crypto/threefish/${MACHINE_ARCH}
+SRCS	+= threefish_512.S
+.else
+SRCS	+= threefish.c
+.endif
 .if exists(${MACHINE_ARCH}/skein_block_asm.s)
 .PATH:	${SRCTOP}/sys/crypto/skein/${MACHINE_ARCH}
 SRCS	+= skein_block_asm.s
