Index: sys/conf/files.mips
===================================================================
--- sys/conf/files.mips	(revision 320320)
+++ sys/conf/files.mips	(working copy)
@@ -86,6 +86,7 @@
 	ipsec_support
 crypto/des/des_enc.c			optional	crypto | ipsec | \
 	ipsec_support | netsmb
+crypto/threefish/threefish.c		optional	crypto | ipsec | ipsec_support
 
 # AP common nvram interface MIPS specific, but maybe should be more generic
 dev/nvram2env/nvram2env.c		optional	nvram2env
