Index: sys/conf/files.riscv
===================================================================
--- sys/conf/files.riscv	(revision 329982)
+++ sys/conf/files.riscv	(working copy)
@@ -5,6 +5,7 @@
 cddl/dev/fbt/riscv/fbt_isa.c				optional dtrace_fbt | dtraceall compile-with "${FBT_C}"
 crypto/blowfish/bf_enc.c	optional	crypto | ipsec | ipsec_support
 crypto/des/des_enc.c		optional	crypto | ipsec | ipsec_support | netsmb
+crypto/threefish/threefish.c	optional	crypto | ipsec | ipsec_support
 dev/ofw/ofw_cpu.c		optional	fdt
 dev/uart/uart_cpu_fdt.c		optional	uart fdt
 dev/xilinx/axi_quad_spi.c	optional	xilinx_spi
