Index: sys/conf/files.mips
===================================================================
--- sys/conf/files.mips	(revision 314509)
+++ sys/conf/files.mips	(working copy)
@@ -84,6 +84,7 @@
 # crypto support -- use generic
 crypto/blowfish/bf_enc.c		optional	crypto | ipsec 
 crypto/des/des_enc.c			optional	crypto | ipsec | netsmb
+crypto/threefish/threefish.c		optional	crypto | ipsec
 
 # AP common nvram interface MIPS specific, but maybe should be more generic
 dev/nvram2env/nvram2env.c		optional	nvram2env
