Index: sys/conf/files.arm64
===================================================================
--- sys/conf/files.arm64	(revision 315705)
+++ sys/conf/files.arm64	(working copy)
@@ -67,6 +67,7 @@
 arm64/cloudabi64/cloudabi64_sysvec.c	optional compat_cloudabi64
 crypto/blowfish/bf_enc.c	optional	crypto | ipsec | ipsec_support
 crypto/des/des_enc.c		optional	crypto | ipsec | ipsec_support | netsmb
+crypto/threefish/threefish.c	optional	crypto | ipsec | ipsec_support
 dev/acpica/acpi_if.m		optional	acpi
 dev/ahci/ahci_generic.c		optional ahci fdt
 dev/hwpmc/hwpmc_arm64.c		optional	hwpmc
