Index: sys/conf/files.amd64
===================================================================
--- sys/conf/files.amd64	(revision 340349)
+++ sys/conf/files.amd64	(working copy)
@@ -183,6 +183,7 @@
 crypto/blowfish/bf_enc.c	optional	crypto | ipsec | ipsec_support
 crypto/des/des_enc.c		optional	crypto | ipsec | \
 	ipsec_support | netsmb
+crypto/threefish/amd64/threefish_512.S optional        crypto | ipsec | ipsec_support
 intel_sha1.o			optional	aesni			\
 	dependency	"$S/crypto/aesni/intel_sha1.c"			\
 	compile-with	"${CC} -c ${CFLAGS:C/^-O2$/-O3/:N-nostdinc} ${WERROR} ${PROF} -mmmx -msse -msse4 -msha ${.IMPSRC}" \
