Index: sys/conf/files.amd64
===================================================================
--- sys/conf/files.amd64	(revision 320320)
+++ sys/conf/files.amd64	(working copy)
@@ -182,6 +182,7 @@
 crypto/blowfish/bf_enc.c	optional	crypto | ipsec | ipsec_support
 crypto/des/des_enc.c		optional	crypto | ipsec | \
 	ipsec_support | netsmb
+crypto/threefish/amd64/threefish_512.S	optional	crypto | ipsec | ipsec_support
 crypto/via/padlock.c		optional	padlock
 crypto/via/padlock_cipher.c	optional	padlock
 crypto/via/padlock_hash.c	optional	padlock
