Index: sys/conf/files
===================================================================
--- sys/conf/files	(revision 313496)
+++ sys/conf/files	(working copy)
@@ -594,6 +594,7 @@
 crypto/skein/skein_block.c	optional crypto | zfs
 crypto/siphash/siphash.c	optional inet | inet6
 crypto/siphash/siphash_test.c	optional inet | inet6
+crypto/threefish/threefish.c	optional crypto | ipsec
 ddb/db_access.c			optional ddb
 ddb/db_break.c			optional ddb
 ddb/db_capture.c		optional ddb
