Index: /usr/src/sys/conf/files.amd64
===================================================================
--- /usr/src/sys/conf/files.amd64	(revision 315075)
+++ /usr/src/sys/conf/files.amd64	(working copy)
@@ -180,6 +180,7 @@
 	no-implicit-rule						\
 	clean		"aesni_wrap.o"
 crypto/blowfish/bf_enc.c	optional	crypto | ipsec
+crypto/threefish/amd64/threefish_512.S	optional	crypto | ipsec
 crypto/des/des_enc.c		optional	crypto | ipsec | netsmb
 crypto/via/padlock.c		optional	padlock
 crypto/via/padlock_cipher.c	optional	padlock
