Index: sys/conf/files.i386
===================================================================
--- sys/conf/files.i386	(revision 315705)
+++ sys/conf/files.i386	(working copy)
@@ -163,6 +163,7 @@
 crypto/via/padlock.c		optional padlock
 crypto/via/padlock_cipher.c	optional padlock
 crypto/via/padlock_hash.c	optional padlock
+cyrpto/threefish/threefish.c	optional crypto | ipsec | ipsec_support
 dev/advansys/adv_isa.c		optional adv isa
 dev/agp/agp_ali.c		optional agp
 dev/agp/agp_amd.c		optional agp
